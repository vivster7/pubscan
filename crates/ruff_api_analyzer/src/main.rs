use std::path::PathBuf;
use std::process::ExitCode;

use anyhow::Result;
use clap::Parser;
use log::debug;

/// CLI arguments for the standalone API analyzer
#[derive(Debug, Parser)]
#[command(
    name = "api-analyzer",
    about = "Analyze a Python module or package to determine its effective public API based on actual usage.",
    version
)]
struct Args {
    /// The path to the Python module (.py file) or package (directory) to analyze.
    #[clap()]
    target: PathBuf,

    /// The output format to use (text/json).
    #[clap(long = "output-format", short = 'o', default_value = "text")]
    output_format: String,

    /// The path to the Python executable to use for venv parsing.
    #[clap(long = "python")]
    python: Option<PathBuf>,

    /// Explicitly specify the project root directory (default: auto-detected from target).
    #[clap(long = "project-root")]
    project_root: Option<PathBuf>,

    /// Disable parallel processing for file analysis.
    #[clap(long)]
    no_parallel: bool,

    /// Increase verbosity (can be used multiple times)
    #[clap(short, long, action = clap::ArgAction::Count)]
    verbose: u8,

    /// Output only a sorted summary line for each symbol.
    #[clap(long)]
    short: bool,
}

fn main() -> ExitCode {
    // Set up colored output
    #[cfg(windows)]
    colored::control::set_virtual_terminal(true).unwrap_or(());

    // Parse command line arguments
    let args = Args::parse();

    // Set up logging based on verbosity
    let log_level = match args.verbose {
        0 => log::LevelFilter::Info,
        1 => log::LevelFilter::Debug,
        _ => log::LevelFilter::Trace,
    };

    env_logger::Builder::new()
        .filter_level(log_level)
        .format_timestamp(None)
        .init();

    // Parse analyzer command using partition
    let analyze_cmd = ruff::args::AnalyzeApiCommand {
        target: args.target.clone(),
        output_format: Some(args.output_format),
        python: args.python,
        project_root: args.project_root,
        preview: false,
        no_preview: false,
        detect_string_imports: false,
        target_version: None,
        no_parallel: args.no_parallel,
        short: args.short,
    };

    // Use Default implementation and rely on ExplicitConfigOverrides for more settings
    let mut global_config = ruff::args::GlobalConfigArgs::default();
    global_config.isolated = true; // Don't try to use .ruff.toml or pyproject.toml
    global_config.config = Vec::new(); // No config options provided

    match analyze_cmd.partition(global_config) {
        Ok((analyze_args, config_args)) => {
            // Call into ruff's analyze_api function
            match run_analyze_api(analyze_args, config_args) {
                Ok(exit_status) => exit_status.into(),
                Err(err) => {
                    eprintln!("Error: {}", err);
                    ExitCode::from(1)
                }
            }
        }
        Err(err) => {
            eprintln!("Error: {}", err);
            ExitCode::from(1)
        }
    }
}

/// Wrapper function to call ruff's analyze_api functionality
fn run_analyze_api(
    args: ruff::args::AnalyzeApiArgs,
    config_args: ruff::args::ConfigArguments,
) -> Result<ruff::ExitStatus> {
    debug!(
        "Running API analysis on target: {}",
        args.target_path.display()
    );
    ruff::commands::analyze_api::analyze_api(&args, &config_args)
}
