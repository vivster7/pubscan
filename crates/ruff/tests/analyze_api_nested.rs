//! Tests the detection of module.symbol patterns in the `analyze api` command.

#![cfg(not(target_arch = "wasm32"))]
#![cfg(not(windows))]

use std::fs;
use std::path::Path;
use std::process::Command;
use std::str;

use anyhow::Result;
use insta_cmd::{assert_cmd_snapshot, get_cargo_bin};
use tempfile::TempDir;

fn command() -> Command {
    let mut command = Command::new(get_cargo_bin("ruff"));
    command.arg("analyze");
    command.arg("api");
    command
}

const INSTA_FILTERS: &[(&str, &str)] = &[
    // Rewrite Windows output to Unix output
    (r"\\", "/"),
    // Redact temporary paths
    (r"/tmp/\.tmp[^/]+", "[TEMPDIR]"),
];

// Create a test package structure with nested modules
fn setup_test_package(root: &Path) -> Result<()> {
    // Create package structure
    fs::create_dir_all(root.join("src").join("pkg1").join("pkg2"))?;

    // Create __init__.py files
    fs::write(root.join("src").join("__init__.py"), "")?;
    fs::write(root.join("src").join("pkg1").join("__init__.py"), "")?;

    // Create the nested module
    fs::write(
        root.join("src")
            .join("pkg1")
            .join("pkg2")
            .join("__init__.py"),
        indoc::indoc! {r#"
        from __future__ import annotations


        def some_function():
            """This is a sample function"""
            return "Hello from pkg2 function"


        class SampleClass:
            """This is a sample class"""

            def method(self):
                return "Hello from pkg2 class method"


        SOME_CONSTANT = "This is a constant"
        "#},
    )?;

    // Create client directory
    fs::create_dir_all(root.join("client"))?;

    // Create client that uses the module.symbol pattern
    fs::write(
        root.join("client").join("__init__.py"),
        indoc::indoc! {r#"
        # Import the module
        from __future__ import annotations

        from src.pkg1 import pkg2

        # Use function via module.symbol pattern
        result = pkg2.some_function()
        print(result)

        # Use class via module.symbol pattern
        instance = pkg2.SampleClass()
        print(instance.method())

        # Use constant via module.symbol pattern
        print(pkg2.SOME_CONSTANT)

        # Also test an if statement that uses the module
        if pkg2.SOME_CONSTANT == "test":
            print("Test logic")
        "#},
    )?;

    Ok(())
}

/// Test detecting module.symbol access patterns
#[test]
fn analyze_module_symbol_access() -> Result<()> {
    let tempdir = TempDir::new()?;
    let root = tempdir.path();

    setup_test_package(root)?;

    // Run the API analysis on the nested module
    insta::with_settings!({
        filters => INSTA_FILTERS.to_vec(),
    }, {
        assert_cmd_snapshot!(
            command()
                .arg(root.join("src").join("pkg1").join("pkg2").join("__init__.py"))
                .current_dir(root)
        );
    });

    Ok(())
}
