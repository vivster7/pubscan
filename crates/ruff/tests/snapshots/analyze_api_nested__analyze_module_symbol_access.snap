---
source: crates/ruff/tests/analyze_api_nested.rs
info:
  program: ruff
  args:
    - analyze
    - api
    - /tmp/.tmp2Tp8qP/src/pkg1/pkg2/__init__.py
---
success: true
exit_code: 0
----- stdout -----
Public API for [TEMPDIR]/src/pkg1/pkg2/__init__.py:

CLASS:
  SampleClass (1 external usages, public)
    Fully qualified: src.pkg1.pkg2.SampleClass
    This is a sample class
    Location: [TEMPDIR]/src/pkg1/pkg2/__init__.py
    Imported by: [TEMPDIR]/client/__init__.py

FUNCTION:
  some_function (1 external usages, public)
    Fully qualified: src.pkg1.pkg2.some_function
    This is a sample function
    Location: [TEMPDIR]/src/pkg1/pkg2/__init__.py
    Imported by: [TEMPDIR]/client/__init__.py

VARIABLE:
  SOME_CONSTANT (2 external usages, public)
    Fully qualified: src.pkg1.pkg2.SOME_CONSTANT
    Location: [TEMPDIR]/src/pkg1/pkg2/__init__.py
    Imported by: [TEMPDIR]/client/__init__.py

Found 3 public API symbols with external usage.

----- stderr -----
[TIMESTAMP][ruff::commands::analyze_api][INFO] Analyzing API for: [TEMPDIR]/src/pkg1/pkg2/__init__.py
[TIMESTAMP][ruff::commands::analyze_api][INFO] Using parallel processing for file analysis
[TIMESTAMP][ruff::commands::analyze_api][INFO] Using parallel processing for file analysis
[TIMESTAMP][ruff::commands::analyze_api][INFO] Auto-detected project root: [TEMPDIR]
